server/src/main/java/com/cloud/api/query/QueryManagerImpl.java

```diff
@@ -2824,75 +2826,42 @@
-         * 'foo%' //sc.setJoinParameters("domainSearch", "path",
-         * domain.getPath() + "%"); // }
-         */
-
-        Pair<List<DiskOfferingJoinVO>, Integer> result = _diskOfferingJoinDao.searchAndCount(sc, searchFilter);
-        // Remove offerings that are not associated with caller's domain
-        if (account.getType() != Account.ACCOUNT_TYPE_ADMIN && CollectionUtils.isNotEmpty(result.first())) {
-            ListIterator<DiskOfferingJoinVO> it = result.first().listIterator();
-            while (it.hasNext()) {
-                DiskOfferingJoinVO offering = it.next();
-                if(!Strings.isNullOrEmpty(offering.getDomainId())) {
-                    boolean toRemove = true;
-                    String[] domainIdsArray = offering.getDomainId().split(",");
-                    for (String domainIdString : domainIdsArray) {
-                        Long dId = Long.valueOf(domainIdString.trim());
-                        if (isRecursive) {
-                            if (_domainDao.isChildDomain(account.getDomainId(), dId)) {
-                                toRemove = false;
-                                break;
-                            }
-                        } else {
-                            if (_domainDao.isChildDomain(dId, account.getDomainId())) {
-                                toRemove = false;
-                                break;
-                            }
-                        }
-                    }
-                    if (toRemove) {
-                        it.remove();
-                    }
-                }
+        // Filter offerings that are not associated with caller's domain
+        // Fetch the offering ids from the details table since theres no smart way to filter them in the join ... yet!
+        Account caller = CallContext.current().getCallingAccount();
+        if (caller.getType() != Account.ACCOUNT_TYPE_ADMIN) {
+            Domain callerDomain = _domainDao.findById(caller.getDomainId());
+            List<Long> domainIds = _domainDao.getDomainParentIds(callerDomain.getId())
+                .stream().collect(Collectors.toList());
+            if (isRecursive) {
+                List<Long> childrenIds = _domainDao.getDomainChildrenIds(callerDomain.getPath());
+                if (childrenIds != null && !childrenIds.isEmpty())
+                domainIds.addAll(childrenIds);
             }
-        }
-        return new Pair<>(result.first(), result.second());
-    }
 
-    private List<ServiceOfferingJoinVO> filterOfferingsOnCurrentTags(List<ServiceOfferingJoinVO> offerings, ServiceOfferingVO currentVmOffering) {
-        if (currentVmOffering == null) {
-            return offerings;
-        }
-        List<String> currentTagsList = StringUtils.csvTagsToList(currentVmOffering.getTags());
+            List<Long> ids = _diskOfferingDetailsDao.findOfferingIdsByDomainIds(domainIds);
+
+            if (ids == null || ids.isEmpty()) {
```

Comment:
This can be refactored as 

```
            SearchBuilder<DiskOfferingJoinVO> sb = _diskOfferingJoinDao.createSearchBuilder();
            if (ids != null && !ids.isEmpty()) {
                sb.and("id", sb.entity().getId(), Op.IN);
            }
            sb.or("domainId", sb.entity().getDomainId(), Op.NULL);
            sb.done();

            SearchCriteria<DiskOfferingJoinVO> scc = sb.create();
            if (ids != null && !ids.isEmpty()) {
                scc.setParameters("id", ids.toArray());
            }
            sc.addAnd("domainId", SearchCriteria.Op.SC, scc);
```